arch/arm64/core/smp.c

```diff
@@ -72,40 +73,49 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 	master_core_mpid = MPIDR_TO_CORE(GET_MPIDR());
 
 	cpu_count = ARRAY_SIZE(cpu_node_list);
+
+#ifdef CONFIG_ARM64_FALLBACK_ON_RESERVED_CORES
+	__ASSERT(cpu_count >= CONFIG_MP_MAX_NUM_CPUS,
+		"The count of CPU Core nodes in dts is not greater or equal to CONFIG_MP_MAX_NUM_CPUS\n");
+#else
 	__ASSERT(cpu_count == CONFIG_MP_MAX_NUM_CPUS,
 		"The count of CPU Cores nodes in dts is not equal to CONFIG_MP_MAX_NUM_CPUS\n");
-
-	for (i = 0, j = 0; i < cpu_count; i++) {
-		if (cpu_node_list[i] == master_core_mpid) {
-			continue;
-		}
-		if (j == cpu_num - 1) {
-			cpu_mpid = cpu_node_list[i];
-			break;
-		}
-		j++;
-	}
-	if (i == cpu_count) {
-		printk("Can't find CPU Core %d from dts and failed to boot it\n", cpu_num);
-		return;
-	}
+#endif
 
 	arm64_cpu_boot_params.sp = Z_KERNEL_STACK_BUFFER(stack) + sz;
 	arm64_cpu_boot_params.fn = fn;
 	arm64_cpu_boot_params.arg = arg;
 	arm64_cpu_boot_params.cpu_num = cpu_num;
 
-	barrier_dsync_fence_full();
+	for (; i < cpu_count; i++) {
+		if (cpu_node_list[i] == master_core_mpid) {
+			continue;
+		}
 
-	/* store mpid last as this is our synchronization point */
-	arm64_cpu_boot_params.mpid = cpu_mpid;
+		cpu_mpid = cpu_node_list[i];
 
-	sys_cache_data_invd_range((void *)&arm64_cpu_boot_params,
-				  sizeof(arm64_cpu_boot_params));
+		barrier_dsync_fence_full();
 
-	if (pm_cpu_on(cpu_mpid, (uint64_t)&__start)) {
-		printk("Failed to boot secondary CPU core %d (MPID:%#llx)\n",
-		       cpu_num, cpu_mpid);
+		/* store mpid last as this is our synchronization point */
+		arm64_cpu_boot_params.mpid = cpu_mpid;
+
+		sys_cache_data_invd_range((void *)&arm64_cpu_boot_params,
+					  sizeof(arm64_cpu_boot_params));
+
+		if (pm_cpu_on(cpu_mpid, (uint64_t)&__start)) {
+			printk("Failed to boot secondary CPU core %d (MPID:%#llx)\n",
+			       cpu_num, cpu_mpid);
+#ifdef CONFIG_ARM64_FALLBACK_ON_RESERVED_CORES
+			continue;
+#else
+			return;
```

Comment:
something is fishy about this logic.

Let's say that you are in the `!CONFIG_ARM64_FALLBACK_ON_RESERVED_CORES` case, you have 5 cores and the 3rd core is refusing to boot.

With the old code you are booting 4 cores out of 5. With your code it looks to me like you are booting the first 2 and then nothing else?

The variable `i` is only being incremented at line `117`, so if you returning here because the core is failing to boot on the next call to the `arch_start_cpu` you are still using the old `i` (static) so you are trying to boot again the core that failed to boot previously?

Reply:
I see what you mean but at this point I think we should then fix `arch_start_cpu()` to return `int` in case the CPU didn't start and gracefully deal with that in `start_cpu()`?

Reply:
I agree it would be good to see the return value propagated down to common code but I believe then it would require a change to all the architecture arch_start_cpu functions.  A much bigger undertaking.  Could we start with this change here, get it in, and then have another PR with the bigger common changes since it will take much more time and affect so many others?